### PR TITLE
FileStatusList: Indicate active file filter and regex error

### LINKED ISSUE #6449: FileStatusList: indicate active filter
## Feature description

An entered file filter for the FileStatusList can be overlooked too easily.
The ComboBox's window color (or text color for a future dark theme) should be changed if the filter string is not empty.
Maybe we could even indicate the validity of the regex, e.g. green and red.

In preparation of switching color schemes, a new Control class should be created.

Thoughts?

### Before

old

### After

regex valid
ok

regex error
error

## Environment

- Git Extensions 3.1.0
- Build 28fcbf1e864ecc2238ef9001586c537957c6c73b
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.2117.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by mstv**
Fixes #6449

## Proposed changes

- indicate active filter of FileStatusList
- indicate regex error in filter of FileStatusList

### Before

old

### After

regex valid
ok

regex error
error

unfiltered
unfiltered focused
unfiltered unfocused

## Test methodology 

- manual tests
- NUnit test for the changes to `FileStatusList.FilterFiles()` factored out as a new method

## Test environment(s) 

- Git Extensions 3.1.0
- Build 0e96050c7ea1bdddf7c98d50ec80529f25e27368
- Git 2.21.0.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.2117.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by mstv** [comment]
Rebased, added tests and minor improvements.

**Comment by RussKie** [comment]
Thank you for the effort. 
I'm not particular happy with theming implementation, I am not sure how it will scale to the whole app and how it affect testability. 
I would like to have a separate discussion about how we can tackle theming in the context of the whole app with the view of possible offering a dark theme.

In an interim, let's just reference/assign colours directly (e.g. `Color.FromArgb(0xFF, 0xC8, 0xC8)`).

Sorry about the confusing directions.

**Comment by mstv** [comment]
I have removed the class `Theme` but kept the branch with the commit c2fb1a50fa1cbf6c2a9b7db815cf1291f383b4b7 as an example for the discussion what should be considered for the dark theme.

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -1518,6 +1538,12 @@ public override int Compare(ListViewItem x, ListViewItem y)
 
         #endregion
 
+        #region private Color constants
+        //// Do not declare the colors "static" because Color.FromArgb() will not work at their initialization.
```
Comment (gerhardol) [inline GitUI/UserControls/FileStatusList.cs]: nitpick: //// is intended for commenting out code (this "feature" is not used now though)

File: GitUI/UserControls/FileStatusList.cs [hunk 2]
```diff
@@ -1518,6 +1538,12 @@ public override int Compare(ListViewItem x, ListViewItem y)
...
+        private readonly Color _activeInputColor = Color.FromArgb(0xC8, 0xFF, 0xC8);
```
Comment (gerhardol) [inline GitUI/UserControls/FileStatusList.cs]: the active color is too visible for my preference but let someone better at UI clean this up...

File: GitUI/UserControls/FileStatusList.cs (hunk 1 above)
Comment (mstv) [reply GitUI/UserControls/FileStatusList.cs]: Yes. Though with only //, the StyleCop scolds "Single-line comment should be preceded by blank line". And I didn't want a blank line after #region here. (The GE code uses both: #region with and without a following blank line.)

File: GitUI/UserControls/FileStatusList.cs (hunk 2 above)
Comment (mstv) [reply GitUI/UserControls/FileStatusList.cs]: It is the same color as used for new lines in the diff view as agreed upon in the feature request.

**Pull request merged**